File: frontend/src/lib/safety.ts
// Safety-layer client — authorization attestations + provenance.
// Active tools refuse to fire against non-lab targets without a stored
// attestation (server-side hard gate); this surfaces + creates them.
import { authFetch } from "../api";

export type Provenance = "lab" | "owned" | "external";

export type Attestation = {
  id: string;
  engagement_id: string | null;
  targets: string[];
  window_start: string;
  window_end: string;
  authority_note: string;
  attested_by: string;
  created_at: string;
};

export async function listAttestations(engagementId: string | null): Promise<Attestation[]> {
  if (!engagementId) return [];
  try {
    const r = await authFetch(`/safety/attestations?engagement_id=${encodeURIComponent(engagementId)}`);
    if (!r.ok) return [];
    const j = await r.json();
    return j.attestations ?? [];
  } catch {
    return [];
  }
}

export async function createAttestation(body: {
  engagement_id: string;
  targets: string[];
  window_start: string;
  window_end: string;
  authority_note: string;
  attested_by: string;
}): Promise<Attestation> {
  const r = await authFetch("/safety/attestations", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return (await r.json()).attestation;
}

export async function getProvenance(target: string): Promise<Provenance> {
  try {
    const r = await authFetch(`/safety/provenance?target=${encodeURIComponent(target)}`);
    if (!r.ok) return "external";
    return (await r.json()).provenance ?? "external";
  } catch {
    return "external";
  }
}

File: backend/routers/safety.py
"""Safety — HTTP surface over `lib/safety.py`.

Mounted at `/safety`. Lets the operator inspect a target's provenance and
manage authorization attestations (the deliberate, gated exceptions that let
an active run reach OUTSIDE the sandbox).

The attestation *gate* itself is enforced server-side inside the active tool
routers (via `lib.scope.enforce_ws/enforce_rest(active=True)`), not here — this
router only creates / lists attestations and reports provenance. There is no
endpoint to bypass the gate; an external active run with no covering
attestation is refused at the tool boundary with a 403.
"""
from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field

from lib import safety
from lib.auth import require_local_auth

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/safety",
    tags=["safety"],
    dependencies=[Depends(require_local_auth)],
)


class AttestationCreate(BaseModel):
    engagement_id: str | None = Field(default=None, max_length=64)
    targets: list[str] = Field(default_factory=list)
    window_start: str = Field(min_length=1, max_length=40)
    window_end: str = Field(min_length=1, max_length=40)
    authority_note: str = Field(default="", max_length=4000)
    attested_by: str = Field(default="", max_length=200)


@router.get("/attestations")
def list_attestations(
    engagement_id: str | None = Query(None, max_length=64),
) -> dict[str, Any]:
    rows = safety.list_attestations(engagement_id)
    return {"count": len(rows), "attestations": rows}


@router.post("/attestations")
def create_attestation(body: AttestationCreate) -> dict[str, Any]:
    att = safety.create_attestation(
        engagement_id=body.engagement_id,
        targets=body.targets,
        window_start=body.window_start,
        window_end=body.window_end,
        authority_note=body.authority_note,
        attested_by=body.attested_by,
    )
    return {"attestation": att}


@router.get("/provenance")
def get_provenance(target: str = Query(..., min_length=1, max_length=2048)) -> dict[str, str]:
    return {"provenance": safety.provenance(target)}
